provider/common/bootstrap.go

```diff
@@ -200,6 +201,31 @@ func BootstrapInstance(ctx environs.BootstrapContext, env environs.Environ, args
 	return result, selectedSeries, finalize, nil
 }
 
+func formatHardware(hw *instance.HardwareCharacteristics) string {
+	if hw == nil {
+		return ""
+	}
+	out := make([]string, 0, 3)
+	if hw.Arch != nil && *hw.Arch != "" {
+		out = append(out, *hw.Arch)
+	}
+	if hw.Mem != nil && *hw.Mem > 0 {
+		out = append(out, fmt.Sprintf("mem=%s", formatMemory(*hw.Mem)))
+	}
+	if hw.CpuCores != nil && *hw.CpuCores > 0 {
+		out = append(out, fmt.Sprintf("cores=%d", *hw.CpuCores))
+	}
+	return strings.Join(out, " ")
+}
+
+func formatMemory(m uint64) string {
```

Comment:
The code you link to is for *parsing* a memory size from a string to an integer. This code is about formatting a memory size (in as an integer in megabytes) to a string expressing either megabytes or gigabytes with one decimal place of precision.

I see your point - and I did consider putting it in a more general location - but I think this function is fairly specialised for this use case (and given the 2.0 release cutoff I don't have time to create something more generic right now).